### ISSUE TITLE
IndexSplitBrainTest.testIndexDoesNotReturnStaleResultsAfterSplit

### ISSUE THREAD

**Issue by hasancelik** [hazelcast/hazelcast#9037]
```
java.lang.AssertionError: entry should be in map2 before split
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at org.junit.Assert.assertNotNull(Assert.java:712)
    at com.hazelcast.query.impl.IndexSplitBrainTest.testIndexDoesNotReturnStaleResultsAfterSplit(IndexSplitBrainTest.java:74)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/1057/testReport/junit/com.hazelcast.query.impl/IndexSplitBrainTest/testIndexDoesNotReturnStaleResultsAfterSplit/

**Comment by hasancelik** [comment]
```
java.lang.AssertionError: Cluster size is not correct. Expected: 2 Actual: 1
	at org.junit.Assert.fail(Assert.java:88)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:750)
	at com.hazelcast.test.HazelcastTestSupport$10.run(HazelcastTestSupport.java:769)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:905)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:765)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:760)
	at com.hazelcast.query.impl.IndexSplitBrainTest.testIndexDoesNotReturnStaleResultsAfterSplit(IndexSplitBrainTest.java:90)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/1083/testReport/junit/com.hazelcast.query.impl/IndexSplitBrainTest/testIndexDoesNotReturnStaleResultsAfterSplit/

**Comment by hasancelik** [comment]
```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds , count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:820)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:813)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:805)
	at com.hazelcast.query.impl.IndexSplitBrainTest.testIndexDoesNotReturnStaleResultsAfterSplit(IndexSplitBrainTest.java:89)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-ZuluJDK6/com.hazelcast$hazelcast/417/testReport/junit/com.hazelcast.query.impl/IndexSplitBrainTest/testIndexDoesNotReturnStaleResultsAfterSplit/

**Comment by mmedenjak** [comment]
The failure for this run
https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/1083/testReport/junit/com.hazelcast.query.impl/IndexSplitBrainTest/testIndexDoesNotReturnStaleResultsAfterSplit/
is because one node was paused for a longer interval (about 2 seconds) after the cluster merged. The cluster clock diff on that node was set to -1893ms and the node did not send any heartbeats for 2 seconds. This caused the node to be removed from the cluster by the master. I can see that the configuration for the test sets the heartbeat timeout to 2 seconds which is far below the default of 300 seconds. I suggest that we increase the heartbeat timeout. @eminn WDYT?

**Comment by mmedenjak** [comment]
In the third test 
https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-ZuluJDK6/com.hazelcast$hazelcast/417/testReport/junit/com.hazelcast.query.impl/IndexSplitBrainTest/testIndexDoesNotReturnStaleResultsAfterSplit/
it looks as if one node had much delays. I approximate that it had about 110 seconds of delays before the test failed. The cluster merge was delayed because of the delayed promotions and the assertion for the cluster merge timed out. There is not much we can do to fix this.

**Comment by mmedenjak** [comment]
@eminn any reasons for setting the heartbeat timeout to 2 seconds? Can we increase it?

**Comment by jerrinot** [comment]
@vbekiaris: can't this be fixed by the change Mehmet and you worked on?

**Comment by vbekiaris** [comment]
@jerrinot yes, the latest failure could be indeed due to the race fixed by #9688
